File: proctoring/app/proctoring/head_pose.py
"""
Head pose estimation and rotation detection
"""
import mediapipe as mp
import numpy as np
from typing import Tuple, Optional

from app.core.logging import logger


class HeadPoseEstimator:
    """Head pose estimation using MediaPipe Face Mesh"""

    # Key face mesh landmarks for head pose
    FACE_LANDMARKS_INDICES = [33, 133, 263, 362, 263, 362, 0, 10, 152, 155, 193, 197]

    def __init__(self):
        """Initialize MediaPipe Face Mesh for head pose"""
        self.mp_face_mesh = mp.solutions.face_mesh
        self.face_mesh = self.mp_face_mesh.FaceMesh(
            static_image_mode=False,
            max_num_faces=1,
            refine_landmarks=True,
            min_detection_confidence=0.7,
            min_tracking_confidence=0.5,
        )

    def estimate_head_pose(
        self,
        image: np.ndarray,
        rotation_threshold: float = 20.0,
    ) -> Tuple[bool, Optional[dict]]:
        """
        Estimate head pose and detect excessive rotation.
        
        Args:
            image: OpenCV image array (BGR)
            rotation_threshold: Threshold in degrees for detecting head turn
            
        Returns:
            Tuple of (is_head_turned, pose_data)
            pose_data contains:
                'pitch': rotation around x-axis (looking up/down)
                'yaw': rotation around y-axis (looking left/right)
                'roll': rotation around z-axis (tilting)
        """
        rgb_image = image[:, :, ::-1]
        results = self.face_mesh.process(rgb_image)

        if not results.multi_face_landmarks or len(results.multi_face_landmarks) == 0:
            return True, None  # No face = head turned away

        landmarks = results.multi_face_landmarks[0].landmark
        h, w, _ = image.shape

        # Use key points for pose estimation
        # Nose
        nose = np.array([landmarks[1].x * w, landmarks[1].y * h])

        # Left and right eye
        left_eye = np.array([landmarks[33].x * w, landmarks[33].y * h])
        right_eye = np.array([landmarks[263].x * w, landmarks[263].y * h])

        # Mouth corners
        left_mouth = np.array([landmarks[61].x * w, landmarks[61].y * h])
        right_mouth = np.array([landmarks[291].x * w, landmarks[291].y * h])

        # Chin
        chin = np.array([landmarks[152].x * w, landmarks[152].y * h])

        # Calculate angles
        # Yaw (left-right rotation)
        eye_distance = right_eye[0] - left_eye[0]
        nose_offset = nose[0] - (left_eye[0] + right_eye[0]) / 2
        yaw = np.degrees(np.arctan2(nose_offset, eye_distance))

        # Pitch (up-down rotation)
        vertical_distance = chin[1] - (left_eye[1] + right_eye[1]) / 2
        nose_vertical_offset = nose[1] - (left_eye[1] + right_eye[1]) / 2
        pitch = np.degrees(np.arctan2(nose_vertical_offset, vertical_distance))

        # Roll (tilting)
        eye_angle = np.degrees(np.arctan2(right_eye[1] - left_eye[1], right_eye[0] - left_eye[0]))
        roll = eye_angle

        # Check if head is turned
        is_head_turned = (abs(yaw) > rotation_threshold or
                         abs(pitch) > rotation_threshold)

        pose_data = {
            "yaw": float(yaw),
            "pitch": float(pitch),
            "roll": float(roll),
        }

        return is_head_turned, pose_data

    def close(self):
        """Close estimator resources"""
        if self.face_mesh:
            self.face_mesh.close()

File: proctoring/src/proctoring/headPose.js
/**
 * Head pose estimation from face landmarks
 * Calculates yaw, pitch, roll from 68-point face landmarks
 */
const logger = require('../config/logger');

// Thresholds for head turn detection (in degrees)
const YAW_THRESHOLD = 25;   // Left/right rotation
const PITCH_THRESHOLD = 20; // Up/down rotation

/**
 * Estimate head pose from face landmarks
 * @param {Buffer} imageBuffer - Image buffer (unused, for API compatibility)
 * @param {Array} landmarks - Array of 68 face landmark points
 * @param {number} rotationThreshold - Threshold in degrees (default: 25)
 * @returns {Object} - {headTurned: boolean, poseData: Object|null}
 */
function estimateHeadPose(imageBuffer, landmarks = null, rotationThreshold = YAW_THRESHOLD) {
  try {
    // If no landmarks provided, we can't estimate pose
    if (!landmarks || landmarks.length === 0) {
      logger.warning('No landmarks for head pose estimation');
      return { headTurned: false, poseData: null };
    }

    // Get first face landmarks if array of arrays
    const faceLandmarks = Array.isArray(landmarks[0]) ? landmarks[0] : landmarks;

    if (faceLandmarks.length < 68) {
      logger.warning(`Insufficient landmarks: ${faceLandmarks.length}`);
      return { headTurned: false, poseData: null };
    }

    // Calculate head pose from key landmarks
    // Using nose tip (30), chin (8), left eye corner (36), right eye corner (45)
    const noseTip = faceLandmarks[30];
    const chin = faceLandmarks[8];
    const leftEye = faceLandmarks[36];
    const rightEye = faceLandmarks[45];
    const leftCheek = faceLandmarks[0];
    const rightCheek = faceLandmarks[16];

    // Calculate yaw (left/right rotation)
    // Asymmetry between nose position and face edges indicates rotation
    const faceWidth = rightCheek.x - leftCheek.x;
    const noseLeftDist = noseTip.x - leftCheek.x;
    const noseRightDist = rightCheek.x - noseTip.x;
    const asymmetry = (noseLeftDist - noseRightDist) / faceWidth;
    const yaw = asymmetry * 90; // Convert to approximate degrees

    // Calculate pitch (up/down rotation)
    // Vertical distance between nose and eyes vs nose and chin
    const eyeCenterY = (leftEye.y + rightEye.y) / 2;
    const noseToEyes = eyeCenterY - noseTip.y;
    const noseToChin = chin.y - noseTip.y;

    // Normal ratio is approximately 0.7
    const normalRatio = 0.7;
    const actualRatio = Math.abs(noseToEyes / noseToChin);
    const pitchOffset = (actualRatio - normalRatio) * 100;
    const pitch = Math.max(-45, Math.min(45, pitchOffset));

    // Calculate roll (head tilt)
    const eyeDiffY = rightEye.y - leftEye.y;
    const eyeDiffX = rightEye.x - leftEye.x;
    const roll = Math.atan2(eyeDiffY, eyeDiffX) * (180 / Math.PI);

    const poseData = {
      yaw: Math.round(yaw * 10) / 10,
      pitch: Math.round(pitch * 10) / 10,
      roll: Math.round(roll * 10) / 10,
    };

    // Check if head is turned beyond threshold
    const headTurned = Math.abs(yaw) > rotationThreshold || Math.abs(pitch) > PITCH_THRESHOLD;

    if (headTurned) {
      logger.info(`Head turned detected: yaw=${poseData.yaw}°, pitch=${poseData.pitch}°`);
    }

    return { headTurned, poseData };
  } catch (error) {
    logger.error(`Head pose estimation error: ${error.message}`);
    return { headTurned: false, poseData: null };
  }
}

/**
 * Check if head is turned away from camera
 * @param {Array} landmarks - Face landmarks
 * @returns {boolean}
 */
function isHeadTurned(landmarks) {
  const result = estimateHeadPose(null, landmarks);
  return result.headTurned;
}

module.exports = {
  estimateHeadPose,
  isHeadTurned,
  YAW_THRESHOLD,
  PITCH_THRESHOLD,
};
